### ISSUE TITLE
Character length for the auto generated code is too long

### ISSUE THREAD

**Issue by navinsubramani** [ni/grpc-labview#17]
The RPC methods and Messages have nested folders, and the auto generated code has more chance to reach the maximum 255 character limit.

**Comment by navinsubramani** [comment]
@ccifra If you have suggestions of what names or folders can be removed or shortened, please comment here. I will look at updating the client and server scripting utility.

Some suggestion from my end:

1. Make the folder name of the class to be "service" instead making the folder with the class name which is too long.
![image](https://user-images.githubusercontent.com/17029551/140843595-7901939b-eda3-4444-9fed-2c5171720f72.png)

2. Make all the VIs under the "RPC Methods" flat instead of nested folders. Given the name of the file is well name spaced, having them flat will not have any conflicts.
![image](https://user-images.githubusercontent.com/17029551/140843757-44c606a4-3d0c-44ac-9571-e9881cab7a4d.png)

3. Make all the VIs under the "RPC messages" flat instead of one nested.
![image](https://user-images.githubusercontent.com/17029551/140843840-b877a3ad-2a50-46d6-817e-59bdf88679b2.png)

Note: This change might or might not support backward compatibility for the scripting tool.

**Comment by KarthikAbiram** [comment]
I agree, making them flat would be a good idea to avoid breaching the 255 length limit. 

One thing to think about is, how would this change affect the previously built servers and is there any easy way to upgrade. Since we are retaining the VI naming and just reorganizing the folders, can a mass compile take care of relinking without any manual work?

**Comment by navinsubramani** [comment]
@KarthikAbiram Good point. Based on my previous discussion with @ccifra , given that this is a not a release product, we are not yet committed to support full backward compatibility. That said, we should define the servers and clients in such a way that they can be scripted with backward support.

I would upload a reference "server template" for sign-off before updating the scripts. The goal is that we align on a reference "server template" that has no character limit challenges, is backward compatible and user code is not disturbed because of scripting.

**Comment by chrisb2244** [comment]
What are the maximum path lengths on different platforms? Is this only a Windows problem?
I found this link [What is LabVIEW File Path Control/Constant Length Limit?](https://knowledge.ni.com/KnowledgeArticleDetails?id=kA00Z0000015AHaSAM) which gives values around what I'd expect (I'm using Win10, and had an error with a path that turned out to be 275 characters) but I'm not sure if the maximum lengths on Mac/Linux/Linux-RT are more forgiving?

I ran into this problem in the "Save VI at Path" VI, but I can see there are a few more locations where files are saved during scripting (e.g the class, library).
Would a change that checked all of these be appropriate, or would targeting specifically the "Save VI at Path" be preferable?

(I'm looking at trying to produce a more descriptive error for this case, given that it seems likely to potentially occur even if this issue is fixed and the paths are flattened in one way or another).

**Comment by KarthikAbiram** [comment]
Checking at all places would be ideal. Checking at Save VI be easier and would cover most common cases.

As per [this stackexchange](https://unix.stackexchange.com/questions/32795/what-is-the-maximum-allowed-filename-and-folder-size-with-ecryptfs),
Linux has a maximum filename length of 255 characters for most filesystems (including EXT4), and a maximum path of 4096 characters. 

While Windows has a path length limit of 255 characters, for Linux it is 4096 characters.

**Comment by navinsubramani** [comment]
[Changes suggested for the template.docx](https://github.com/ni/grpc-labview/files/7572269/Changes.suggested.for.the.template.docx)

Please find the list of changes suggestion and the t
[service.zip](https://github.com/ni/grpc-labview/files/7572271/service.zip)
emplate service which has those changes in them.

**Comment by twaltersp** [comment]
We have a complicated proto file (actually a library of them with includes) that we can't support due to path names.  The error we are running into currently is related to creating the message data clusters.  In this example the proto has a message with a number of other embedded messages.  This leads LV to try to use the name: simian_public_sensor_model_Description_RadarModel_SensorFidelity_RaytracingConfig_AdaptiveSamplingParams_Target.

I've already tried to make the build path as short at possible but I can't modify the proto files since they are from a third party.

**Comment by twaltersp** [comment]
I think the approach of directly using the names pulled from the proto files will be a challenge unless LV changes to support the full path that the OS provides (LV is limited to 256 char total path length I think).  

Perhaps the scripted code approach needs to change to not use the direct name and instead have a map or other lookup function?

**Comment by bkeryan** [comment]
Some more feedback on this issue based on encountering path length limits with other gRPC-related projects:

By default, Windows has a path limit of 260 characters. This is documented in [Maximum Path Length Limitation](https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry). There are two ways to bypass this limit:
- Use the Unicode versions of Win32 APIs (e.g. `CreateFileW` instead of `CreateFileA`) and specify a `\\?\` path prefix. I think recent versions of the .NET runtime do this automatically, but doing it in C/C++ applications like LabVIEW would require a lot of care and may touch a lot of code.
- Use Windows 10 version 1607 or later, configure the system-wide registry setting to enable long paths, and update the application manifest to opt-in to long paths. This manifest is embedded into the application EXE, such as `LabVIEW.exe` or `python.exe`.

I think Microsoft made long path functionality opt-in because enabling long paths for applications that use hardcoded arrays like `char filename[PATH_MAX]` can make the application crash and this is often a security vulnerability.

I have run into path length issues many times when working on NI's [measurement-plugin-python](https://github.com/ni/measurement-plugin-python) project, caused by that project's Python protobuf codegen. The manifest for `python.exe` opts in to long path support, and the standard Python installer asks if you want to configure the system-wide registry setting to enable long paths. However, alternative installation methods like pyenv-win do not configure the system-wide registry setting. If you install Python via pyenv-win and your virtualenv is located in a longer-than-usual directory such as `C:\Users\Administrator\Downloads\measurement-plugin-python-examples-1.2.3\measurement-plugin-python\examples\some_measurement\...` you may run into problems. That's why we document how to configure the system-wide registry setting in the readme for that project: [System Configuration >> Enable Win32 Long Paths](https://github.com/ni/measurement-plugin-python?tab=readme-ov-file#enable-win32-long-paths).

I haven't spent as much time with the corresponding [measurement-plugin-labview](https://github.com/ni/measurement-plugin-labview/) project, but I think the same protos that cause issues for measurement-plugin-python provide a good example of this issue. [Source/gRPC/Generated APIs/ni/measurementlink/sessionmanagement/v1/RPC Messages/FlatToRichni_measurementlink_sessionmanagement_v1_ReserveSessionsResponse_GroupMappingsEntry.vi](https://github.com/ni/measurement-plugin-labview/blob/main/Source/gRPC/Generated%20APIs/ni/measurementlink/sessionmanagement/v1/RPC%20Messages/FlatToRichni_measurementlink_sessionmanagement_v1_ReserveSessionsResponse_GroupMappingsEntry.vi) is about 175 characters. This leaves users with only about 85 characters for the driver letter and the path to the measurement-plugin-labview codebase. Something like `C:\Program Files\National Instruments\LabVIEW 2024\user.lib` or `C:\Users\Administrator\Downloads\measurement-plugin-labview` is about 60 characters, so it's probably not hard to hit this by accident.

When I look at the path [Source/gRPC/Generated APIs/ni/measurementlink/sessionmanagement/v1/RPC Messages/FlatToRichni_measurementlink_sessionmanagement_v1_ReserveSessionsResponse_GroupMappingsEntry.vi](https://github.com/ni/measurement-plugin-labview/blob/main/Source/gRPC/Generated%20APIs/ni/measurementlink/sessionmanagement/v1/RPC%20Messages/FlatToRichni_measurementlink_sessionmanagement_v1_ReserveSessionsResponse_GroupMappingsEntry.vi), one thing that jumps out at me is that the gRPC package name `ni.measurementlink.sessionmanagement.v1` is duplicated between the directory path and the VI filename. LabVIEW requires VI and ctl filenames to be unique, unless they are owned by a `.lvlib`, in which case the `.lvlib` is used to namespace the VI and ctl filenames. Since `ni.measurementlink.sessionmanagement.v1.api.lvlib` contains the gRPC package name, then I think the VI and ctl filenames should not need to contain the gRPC package name.

`ReserveSessionsResponse_GroupMappingsEntry` is a nested message that is automatically generated for the map field. I don't think this is avoidable, but it's a reason to avoid overly long field names, at least for maps.

`simian_public_sensor_model_Description_RadarModel_SensorFidelity_RaytracingConfig_AdaptiveSamplingParams_Target` sounds like it makes heavy use of nested messages. I don't know of a good way to shorten these paths other than mapping them to user-chosen names, as suggested. Note that some protobuf style guides, such as the [Buf Style Guide](https://buf.build/docs/best-practices/style-guide/#recommendations), discourage use of nested messages, but that's no consolation if you're trying to call or implement an existing API that uses nested messages.
